ogre2/src/Ogre2SegmentationCamera.cc

```diff
@@ -0,0 +1,647 @@
+            float instanceColor2 = (instanceCount % 256) / 255.0;
+
+            subItem->setCustomParameter(1, Ogre::Vector4(
+              labelColor, instanceColor1, instanceColor2, 1.0));
+          }
+        }
+
+        // check if it's an overlay material by assuming the
+        // depth check and depth write properties are off.
+        if (!datablock->getMacroblock()->mDepthWrite &&
+            !datablock->getMacroblock()->mDepthCheck)
+          subItem->setMaterial(this->plainOverlayMaterial);
+        else
+          subItem->setMaterial(this->plainMaterial);
+      }
+    }
+    itor.moveNext();
+  }
+
+  // reset the count & colors tracking
+  this->instancesCount.clear();
+  this->takenColors.clear();
+  this->coloredLabel.clear();
+}
+
+/////////////////////////////////////////////////
+void SegmentationMaterialSwitcher::postRenderTargetUpdate(
+    const Ogre::RenderTargetEvent &/*_evt*/)
+{
+  // restore item to use hlms material
+  auto itor = this->scene->OgreSceneManager()->getMovableObjectIterator(
+      Ogre::ItemFactory::FACTORY_TYPE_NAME);
+  while (itor.hasMoreElements())
+  {
+    Ogre::MovableObject *object = itor.peekNext();
+    Ogre::Item *item = static_cast<Ogre::Item *>(object);
+    for (unsigned int i = 0; i < item->getNumSubItems(); i++)
+    {
+      Ogre::SubItem *subItem = item->getSubItem(i);
+      auto it = this->datablockMap.find(subItem);
+      if (it != this->datablockMap.end())
+        subItem->setDatablock(it->second);
+    }
+    itor.moveNext();
+  }
+}
+
+/////////////////////////////////////////////////
+class ignition::rendering::Ogre2SegmentationCameraPrivate
+{
+  /// \brief Material Switcher to switch item's material
+  /// with colored version for segmentation
+  public: SegmentationMaterialSwitcher *materialSwitcher {nullptr};
+
+  /// \brief Compositor Manager to create workspace
+  public: Ogre::CompositorManager2 *ogreCompositorManager {nullptr};
+
+  /// \brief Workspace to interface with render texture
+  public: Ogre::CompositorWorkspace *ogreCompositorWorkspace {nullptr};
+
+  /// \brief Workspace Definition
+  public: std::string workspaceDefinition;
+
+  /// \brief Render Texture to store the final segmentation data
+  public: Ogre::RenderTexture *ogreRenderTexture {nullptr};
+
+  /// \brief Texture to create the render texture from.
+  public: Ogre::TexturePtr ogreTexture;
+
+  /// \brief Pixel Box to copy render texture data to a buffer
+  public: Ogre::PixelBox *pixelBox {nullptr};
+
+  /// \brief buffer to store render texture data & to be sent to listeners
+  public: uint8_t *buffer = nullptr;
+
+  /// \brief dummy render texture to set image dims
+  public: Ogre2RenderTexturePtr dummyTexture {nullptr};
+
+  /// \brief New Segmentation Frame Event to notify listeners with new data
+  public: ignition::common::EventT<void(const uint8_t *, unsigned int,
+        unsigned int, unsigned int, const std::string &)>
+        newSegmentationFrame;
```

Comment:
Can you document the template parameter of this variable? It looks like it's a callback method with a few arguments, so don't forget to document the arguments to the callback as well :slightly_smiling_face: